File: packages/twerk-slots/src/twerk_slots/cli/slot/gt/up.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated

import click

from twerk_core.clinkr.exit import ClinkrExit
from twerk_core.clinkr.operation import clinkr_operation
from twerk_core.git.types import DetachedHead
from twerk_core.git.types import GitCommandFailure as GitFailure
from twerk_core.gt.types import GtCommandFailure, UntrackedBranch
from twerk_slots.cli.slot.gt.context import load_slot_gt_context
from twerk_slots.cli.slot.gt.navigation import (
    GtNavigationTarget,
    build_navigation_result,
    find_worktree_for_branch,
    render_gt_navigation,
)
from twerk_slots.repo_context import NoRepoSentinel


@dataclass(frozen=True)
class SlotGtUpRequest:
    no_clipboard: Annotated[
        bool,
        click.Option(["--no-clipboard"], is_flag=True, default=False),
    ] = False


@clinkr_operation(
    name="up",
    help="Print the worktree for the immediate upstack Graphite branch.",
    human_renderer=render_gt_navigation,
)
def run_gt_up(ctx: click.Context, request: SlotGtUpRequest) -> ClinkrExit[GtNavigationTarget]:
    gt_ctx = load_slot_gt_context(ctx)
    if isinstance(gt_ctx, NoRepoSentinel):
        raise ClinkrExit.failure(error_type="not_in_repo", message=gt_ctx.message)

    slots_ctx = gt_ctx.slots
    current_branch = slots_ctx.git.get_current_branch(slots_ctx.repo.root)
    if isinstance(current_branch, GitFailure):
        raise ClinkrExit.failure(
            error_type="git_current_branch_failed",
            message=current_branch.message,
        )
    if isinstance(current_branch, DetachedHead):
        raise ClinkrExit.failure(
            error_type="detached_head",
            message=f"HEAD at {slots_ctx.repo.root} is detached. Check out a branch first.",
        )
    current = current_branch

    children = gt_ctx.gt.children_of(slots_ctx.repo.root)
    if isinstance(children, UntrackedBranch):
        raise ClinkrExit.failure(
            error_type="untracked_branch",
            message=f"Current branch '{current}' is not tracked by Graphite. {children.message}",
        )
    if isinstance(children, GtCommandFailure):
        raise ClinkrExit.failure(error_type="gt_children_failed", message=children.message)
    if isinstance(children, tuple) and len(children) == 0:
        raise ClinkrExit.negative(message=f"No upstack branch for '{current}'.")
    if isinstance(children, tuple) and len(children) > 1:
        candidates = ", ".join(children)
        raise ClinkrExit.negative(
            message=(
                f"Multiple upstack branches for '{current}': {candidates}. "
                "Run `slot checkout <branch>` for the branch you want."
            )
        )

    (child,) = children

    target = find_worktree_for_branch(slots_ctx, child)
    if target is None:
        raise ClinkrExit.negative(
            message=(
                f"Upstack branch '{child}' is not checked out in any worktree. "
                f"Run `slot checkout {child}`."
            )
        )
    return ClinkrExit.ok(build_navigation_result(slots_ctx, target, request.no_clipboard))

File: ts/packages/slot/src/operations/gt/up.ts
import { failure, negative, ok } from "@asdl/clinkr";
import { z } from "zod";

import type { SlotCliContext } from "../../context.ts";
import { buildGtNavigationResult, renderGtNavigation, resolveOrCheckoutWorktreeForBranch } from "./navigation.ts";
import { resolveRepoAndCurrentBranch } from "./shared.ts";

export const gtUpRequestSchema = z.object({
	clipboard: z.boolean().default(true).describe("Copy the cd command to the clipboard."),
});

export const gtNavigationResultSchema = z.object({
	slot_name: z.string().nullable(),
	branch_name: z.string(),
	worktree_path: z.string(),
	cd_command: z.string(),
	is_already_assigned: z.boolean(),
	is_clipboard_copied: z.boolean(),
	is_clipboard_skipped: z.boolean(),
	clipboard_failure_reason: z.union([z.literal("backend_missing"), z.literal("subprocess_error")]).nullable(),
	clipboard_failure_detail: z.string().nullable(),
});

export type GtUpRequest = z.infer<typeof gtUpRequestSchema>;

export async function runGtUp(ctx: SlotCliContext, request: GtUpRequest) {
	const resolved = await resolveRepoAndCurrentBranch(ctx);
	if (resolved.type !== "ok") return resolved;
	const children = await ctx.gt.childrenOf(resolved.repoCtx.repo.root);
	if (children.type === "untracked_branch") return failure("untracked_branch", `Current branch '${resolved.currentBranch}' is not tracked by Graphite. ${children.message}`);
	if (children.type === "failure") return failure("gt_children_failed", children.failure.message);
	if (children.branches.length === 0) return negative(`No upstack branch for '${resolved.currentBranch}'.`);
	if (children.branches.length > 1) return negative(`Multiple upstack branches for '${resolved.currentBranch}': ${children.branches.join(", ")}. Run \`slot checkout <branch>\` for the branch you want.`);
	const branch = children.branches[0] ?? "";
	const resolution = await resolveOrCheckoutWorktreeForBranch(ctx, branch);
	if (resolution.type === "failure") return resolution;
	return ok(await buildGtNavigationResult(ctx, resolution.resolution, { shouldSkipClipboard: !request.clipboard }));
}

export { renderGtNavigation };
